### ISSUE TITLE
eyre: connection requires frequent refresh to maintain

### ISSUE THREAD

**Issue by jalehman** [urbit/urbit#6311]
Users frequently have to refresh the page in order to maintain a connection to eyre. We don't really know why.

@dr-frmr @tadad and @drunkplato are actively looking into this and/or have expressed interest in seeing it resolved. If you have any further context or findings to add, please do so here.

**Comment by tadad** [comment]
so the source of the bug seems to be any time that you don't close a session, but the internet goes out. So sleeping your laptop, backgrounding an app in ios, disconnecting from wifi, etc. all trigger this. @e1block and I have a way to easily reproduce the bug 

First step is adding some extra printouts to %eyre, so finishing up #6303 to help with bug hunting.

**Comment by matthew-levan** [comment]
As far as I understand it (I'm not very familiar with how server-sent events work), the correct flow is something like this:
- A session's underlying internet connection is broken
- Internet connection is later restored
- Client re-establishes an initial connection with the server
- Server continues sending data over the re-established connection

**Comment by belisarius222** [comment]
As @Fang- said, this sounds like it could be a purely client-side issue.  Is there an industry best-practice for handling connectivity loss for a client using server-sent events?  Seems like something that would affect anyone trying to use them in production.

A cursory reading of some SO posts about SSS indicate the browser should automatically try to reconnect.  Not sure if Chrome is just failing here, or if we have something configured awkwardly (a timeout interval set too long, perhaps), or what.

Possibly relevant: https://stackoverflow.com/a/58374643

**Comment by matthew-levan** [comment]
I think we might need to produce reliable reproduction steps for this bug before we can make much more progress. FWIW, I can't reproduce the "forever reconnecting" behavior in Landscape by switching off my network adapter, scrolling far up enough in a chat for Landscape to send a request to the server for more data (and thereby rendering the yellow "Reconnecting..." bar on top of the screen), and then re-enabling my network adapter. In fact, Landscape proceeds to reconnect in this scenario, succeeds, and loads the data it was asking for just fine.

**Comment by tadad** [comment]
<img width="766" alt="Captura de Pantalla 2023-02-13 a la(s) 12 55 12 p  m" src="https://user-images.githubusercontent.com/43223192/218548966-7d6590fe-6b11-4708-8a57-9b9a7297d887.png">
something is really weird here - so the issue stems not from disconnecting - but reconnecting. Somehow on reconnect, something is passing a %cancel-request event directly into %eyre. Must be coming from the frontend? Need to look more

**Comment by arthyn** [comment]
we've taken to doing our own type of heartbeat from the frontend to check the channel's status by constantly sending a poke to settings-store and listening for the fact coming back. tloncorp/landscape-apps#1886

**Comment by tadad** [comment]
interesting. I'm pretty sure something properly implemented wouldn't require a client-side heartbeat system.

I've narrowed down the problem to *something* sending a `[%cancel-request ~]` event into %eyre, which means that %eyre seems to be doing it's job correctly - i.e. something tells it to disconnect and it does. However, whatever is telling it to disconnect (the frontend probably) *could* in theory keep the connection alive, it's just killing it prematurely

**Comment by arthyn** [comment]
Yeah I still think we should fix what's going on, just wanted to share what we're doing until then.

Tracing through eyre and http-api a little bit, I think what you're looking for is this line here: https://github.com/urbit/urbit/blob/5c73e70c2ff0c0ec9be5cd12b93f050049f7f095/pkg/npm/http-api/src/Urbit.ts#L328

Can't find where this translates into %cancel-request but you can see the comment here: https://github.com/urbit/urbit/blob/5c73e70c2ff0c0ec9be5cd12b93f050049f7f095/pkg/arvo/sys/vane/eyre.hoon#L850

**Comment by pkova** [comment]
It seems to me that the problem is here:

https://github.com/urbit/urbit/blob/5c73e70c2ff0c0ec9be5cd12b93f050049f7f095/pkg/npm/http-api/src/Urbit.ts#L304-L306

It retries five times and then gives up, probably not the behavior we want.

**Comment by tadad** [comment]
@e1block and I dug into this more and have some new findings:

There is definitely merit to what @pkova  pointed out, it shouldn’t be cancelling after a set number of tries. However, we don’t believe this is the only bug: 
We believe that you can fix this by using the onRetry and on onError correctly - which not many (any?) ui repos do
- `onError` should continue propagate error messages rather than just stopping.
- `onError` you should also retry the connection (`onRetry`) - most likely using some sort of exponential backoff delay for retries. We believe we can implement an auto reconnect right here.
- Also since it seems like no one knows how to use `onRetry` and `onError` correctly, we should just write a robust version once and get rid of it from the http-api interface, rather than having each frontend app implement their own

Could still be a deep network stack error that we will have to fix (see [here](https://github.com/Azure/fetch-event-source/pull/27), no idea why this issue was closed).

**Comment by tadad** [comment]
going to offer @liam-fitzgerald to chime in since [he wrote a lot of this](https://github.com/urbit/urbit/blame/develop/pkg/npm/http-api/src/Urbit.ts)

**Comment by tadad** [comment]
starting to think the @microsoft/asdfasdfad dependency is the problem: in [the readme](https://github.com/Azure/fetch-event-source#fetch-event-source) it literally says: "If the connection is cut, you don't have any control over the retry strategy: the browser will silently retry for you a few times and then stop, which is not good enough for any sort of robust application."

**Comment by tadad** [comment]
Look at this comment : https://github.com/Azure/fetch-event-source/blob/main/src/fetch.ts#L45 (logic that implements this is line 78)
***This is the reconnect bug***. Have to ditch this dependency and write it from scratch

**Comment by tacryt-socryp** [comment]
Yup, that `.abort()` on line 76 will cancel the Eyre channel. Great discovery.

**Comment by matildepark** [comment]
> starting to think the @microsoft/fetch-event-source dependency is the problem: in [the readme](https://github.com/Azure/fetch-event-source#fetch-event-source) it literally says: "If the connection is cut, you don't have any control over the retry strategy: the browser will silently retry for you a few times and then stop, which is not good enough for any sort of robust application."

That line is referring to the problem it is trying to address with the package.



> Look at this comment : https://github.com/Azure/fetch-event-source/blob/main/src/fetch.ts#L45 (logic that implements this is line 78)
> _**This is the reconnect bug**_. Have to ditch this dependency and write it from scratch

We already set that to `true` here:

https://github.com/urbit/urbit/blob/a9b4c220dd219b30a8bcc67336f4fb7ea0769f95/pkg/npm/http-api/src/Urbit.ts#L220

**Comment by tadad** [comment]
yeah got ahead of myself, that's not it

**Comment by tadad** [comment]
had some more time to think about the problem, I'm convinced we can't get a good UX in the browser without ripping out SSE from `%eyre`. More details here: https://dachus-tiprel.tlon.network/proposal/websockets

**Comment by tadad** [comment]
since websockets are a large project, I'm going to implement some quick fixes first, read about them here: https://dachus-tiprel.tlon.network/proposal/eyre-fix comments welcome. Main idea is to get rid of all subscriptions ASAP and use polling

**Comment by pkova** [comment]
This is how I see all of this.

When the socket is killed cleanly things are easy, but a dirty kill (pull the network cable, suspend, who knows what) is the trickier case. A clean kill is easy to simulate, just shut down your ship normally. A dirty kill can also be simulated by sending the urbit processes a `SIGSTOP`
```
$ pgrep urbit
51693
51694
$ kill -s STOP 51594 51593
```
This is where the SSE client runs into problems, since it doesn't error out for many minutes. Note that we can resume by doing `kill -s CONT 51594 51593`, and remember to `fg` whatever terminal window you were running the ship on.

An obvious fix presents itself since Eyre is sending a SSE heartbeat every 20 seconds: just error out if don't hear anything in 25 seconds! Both of these values are of course tunable.

Annoyingly, the SSE protocol itself doesn't let us access the heartbeats, so we have to muck with the implementation which is currently `@microsoft/fetch-event-source`.

I have created [a repo](https://github.com/pkova/reconnect-test) that includes a Gall app and all the relevant frontend pieces  (`@microsoft/fetch-event-source`, `@urbit/http-api`) in one codebase to make testing the SSE connection easy. I made the following changes to the frontend pieces:

- Instead of `@urbit/http-api` retrying five times and giving up it keeps retrying forever every 5 seconds.
- Implement `responseTimeout` in `@microsoft/fetch-event-source`, throwing if getting bytes takes longer than the timeout. 25 seconds is used in the repo.
- Implement `onreconnect` in `@microsoft/fetch-event-source` so we can detect when a reconnect actually happens.

With these changes I have been unable to find a case where the reconnect fails in my test application. I would be interested to know if anybody else can break it.

**Comment by arthyn** [comment]
well you beat me to it haha! I was starting down the same path but didn't make it as far as you: https://github.com/tloncorp/eyrie

I'll look through this and see what I can find. We probably should just ship our own version of fetch-event-source wrapped into http-api anyway since it seems somewhat unmaintained. Thank you for looking deeper into this.

**Comment by belisarius222** [comment]
@pkova Can you make a PR with your changes in urbit/urbit, since that's the canonical source for the http-api?  I wonder if we should just vendor your forked @microsoft/fetch-event-source library for the time being.

**Comment by arthyn** [comment]
probably should go here https://github.com/urbit/js-http-api since we're trying to move it out of urbit/urbit

**Comment by ajlamarc** [comment]
> well you beat me to it haha! I was starting down the same path but didn't make it as far as you: https://github.com/tloncorp/eyrie
> 
> I'll look through this and see what I can find. We probably should just ship our own version of fetch-event-source wrapped into http-api anyway since it seems somewhat unmaintained. Thank you for looking deeper into this.

Can it be based on the `@fortaine/microsoft-event-source` fork so it also works in node context? https://github.com/urbit/js-http-api/pull/1  (discovered by @pkova as well)

**Comment by arthyn** [comment]
yeah let's do this. let's get your PR merged and then I'll diff @pkova 's changes onto it

**Comment by e1block** [comment]
Update from the iOS side.

* As mentioned by @tadad, I added logic client side for TCP re-connection on the app returning to foreground from background
* This works consistently, however, while in the background, we're not ACK'ing events. If too many go un-ACK'ed (~30?), the connection is killed. However, Eyre debug reports the channel as alive and no un-ACK'ed events. See `GET /~debug/eyre/channels.json` or `/~debug/eyre` in Landscape. Can send pokes, but no events come through. Found this while debugging with @leowini.

![Screenshot 2023-04-14 at 4 39 05 PM](https://user-images.githubusercontent.com/104395039/232541031-17ff3701-0d4b-44d7-9666-e8b1ab0f170e.png)

* We can add logic to ACK events even while in the background. However, you could still lose signal for a time and be unable. Point here is, we need a reliable way to know the channel is dead or not.
* Furthermore, with clients such as mobile, the use case for SSE is different. You only have _**one**_ instance of the client, and thus only one SSE channel. In the case of the browser, you could have X number of Landscape tabs. So the idea of constantly killing and creating channels makes more sense in that context. For mobile, we'd want basically a single channel that can **ALWAYS** be resumed.

**Comment by joemfb** [comment]
Even with the mobile use-case of a single, long-lived channel, it still needs to be killed to avoid a space leak if un-ack'd events pile up.

**Comment by e1block** [comment]
> Even with the mobile use-case of a single, long-lived channel, it still needs to be killed to avoid a space leak if un-ack'd events pile up.

@joemfb: For sure. Since were pausing adding events to the queue/channel after X time, and the client will have to scry to get the delta of events while away, in addition to create & sub to a new channel, my suggestion is to allow re-connecting on that channel, which will send new events from that point onward, without space leak. Essentially the case that's happening above.

Put another way, why can't we keep using the channel above? Why force the client to _"restart"_?

**Comment by joemfb** [comment]
If the channel is not killed but only paused, an arbitrary number of events from a given subscription could be missed. Some applications might be fine with "sparse subscription updates", but in general, that would break the abstraction.

I'm not sure what you mean about "scry to get the delta", how would that delta be identified and addressed?

**Comment by e1block** [comment]
> Some applications might be fine with "sparse subscription updates", but in general, that would break the abstraction.

@joemfb: Indeed. As far as I've seen, agents don't seem to have that logic (built-in default). Are we supposed to be able to pass a timestamp on a subscription to get all events missed since timestamp X? (when creating a new channel). Were doing below instead.

> I'm not sure what you mean about "scry to get the delta", how would that delta be identified and addressed?

So lets say in the case of chat. I miss several messages while away, and the events never get to me on the channel for whatever reason. I'm going to have to scry on the DM path to get everything I've missed (the delta - whole DM log technically), and then also need a connection for real time events/messages (I have the app open and active now).

In our particular case, were passing a timestamp to the agent on scry to only send what we've missed, but that's a case specific optimization.

**Comment by e1block** [comment]
> how would you find out that you need to synchronize (scry for dm backlog, for example)?

@joemfb: Anytime we open a DM, we scry by default. Anytime, you return from background, we scry the DM list. This is standard mobile app behavior (opening a Twitter profile or IG feed for example).

Since chats are cached locally we present that while we scry on a background thread, so its seamless from a user perspective (UI is not blocked), and prevents the risk of anything missing because of dropped events. It's also explicit, because there's a clear request/response flow, rather than waiting for unknown # of events to come back on the channel (that's been newly made).

> In the current model, killing the channel serves this purpose.

How does the client know the channel is dead? We don't get an indication as far as I can tell (error or event) - https://github.com/urbit/urbit/issues/6311#issuecomment-1511653452

**Comment by joemfb** [comment]
> How does the client know the channel is dead?

That's a good question, I believe clients need to send heartbeats. @Fang- @pkova @arthyn?

**Comment by arthyn** [comment]
I think we need some diagrams with documentation because I think things are being in lost in translation both here and elsewhere. I'll try to get some going w/ @pkova and map out the behavior so people can see clearly.

**Comment by e1block** [comment]
Spoke with @pkova & @arthyn yesterday.

With regards to this instance (https://github.com/urbit/urbit/issues/6311#issuecomment-1511653452), there is a quit event sent on TCP re-connect indicating the subscription is finished. The channel is still valid & active. Hence, `/~debug/eyre/channels.json` will return the channel with an empty subscription list. Sending a subscription request on the same channel should resume it (haven't tested yet). @pkova's new PR (#6490) addresses the issue with the queued events being dropped. After 12 hours of inactivity however, the channel itself will be killed. Returns a `404` on connection re-attempt (haven't tested this yet).
